### PR TITLE
feat(elf): discover maximal supported optimization windows

### LINKED ISSUE #617: --auto driver: window-finder (maximal supported straight-line runs)
Part of #615. Implements **ADR-0009 Decision 3** (and the v1 RIP-relative exclusion of Decision 6).

## Background

The driver needs to discover its own candidate windows instead of taking
`--start-addr`/`--end-addr`. A candidate window is a **maximal run of
straight-line instructions that the existing search can actually handle**.
"Supported" is defined operationally by the existing Capstone→IR bridge: a
window is admissible only if every instruction in it converts through
`convert_capstone_op_for_optimization` (`src/main.rs`) without rejection. We do
**not** maintain a second mnemonic allow-list (mirrors the
parser-is-single-source-of-truth rule in `CLAUDE.md`), so the finder cannot
drift from the search.

## Scope

- Walk each executable section's Capstone disassembly (reuse existing Capstone
  setup and the per-arch `convert_to_ir` / `convert_capstone_op_for_optimization`
  path).
- Emit candidate windows = maximal runs where **every** instruction converts to
  IR successfully.
- **Cut** a window at: an unsupported instruction, a `call`, or a control-flow
  terminator (terminators may sit at the boundary held fixed but never in the
  interior).
- **Exclude** windows containing a RIP-relative memory operand (ADR-0009
  Decision 6 — displacements are next-IP-relative and the x86 IR does not model
  them yet).
- Output: a list of `(start_addr, end_addr)` candidates per section.

## Out of scope

- Branch-target exclusion → slice for **direct targets** and the **indirect/jump
  table** soundness slice (both tracked under #615).
- The loop, prioritization, patching.

## Acceptance criteria

- Unit tests on small byte fixtures: a supported run flanked by unsupported
  instructions yields exactly the inner window bounds.
- A `call` or terminator never appears in a window interior.
- A window containing a RIP-relative operand is excluded.

## Code pointers

- `convert_to_ir` / `convert_capstone_op_for_optimization` — `src/main.rs`.
- `ElfPatcher` section/address access — `src/elf_patcher/`.
- Per-arch Capstone construction used by `disasm`/`opt`.

### PR DESCRIPTION
**Pull request by pmatos**
## Summary

- discover end-exclusive candidate windows for every non-empty executable ELF section
- derive support from the existing AArch64 and x86 Capstone-to-IR conversion paths
- split at unsupported instructions and calls, hold supported terminators only at trailing boundaries, and exclude x86-64 RIP-relative memory operands
- fail closed on missing instruction detail and partial section disassembly

## Verification

- `cargo fmt --all -- --check`
- `cargo clippy --all -- -D warnings`
- `cargo test --all`
- `./ci_check.sh`

Closes #617

**Merge with `gh pr merge --merge` (merge commit, not squash) per CLAUDE.md.**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -3929,10 +4234,231 @@
...
+                rm: Operand::Immediate(1),
+            },
+        ]);
+        let cs = aarch64_test_capstone();
+        let disassembly = cs
+            .disasm_all(&bytes, 0x1000)
+            .expect("fixture should disassemble");
+        let call = disassembly
+            .iter()
+            .nth(1)
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main.rs]: Fixed in `c313223`.

Changed: `candidate_windows_exclude_calls_and_split_both_sides` now uses `disassembly.get(1)` instead of `disassembly.iter().nth(1)`. `Instructions` derefs to `[Insn]`, so `.get(1)` yields the same `Option<&Insn>` — a semantics-preserving drop-in that clears the clippy `iter_nth` finding.
Validation: `cargo clippy --all-targets --all-features` (the `iter_nth` finding is gone, no new `main.rs` warnings), `cargo fmt --all -- --check`, and the `candidate_windows` test suite all pass.

**Comment by pmatos** [comment]
@claude Regarding your review (*discover maximal supported optimization windows*) — triaged the three actionable suggestions with two independent evaluators (Claude Opus 4.8 and Codex GPT-5.x):

**Test-coverage gap → Fixed in `157e32a`.**
Added `candidate_windows_split_run_at_unsupported_instruction`: an unsupported `push rax` between two supported runs (`add rax,1; push rax; sub rbx,1`) now exercises the `Err(_)` flush branch and produces two windows (`[0x1000,0x1004]` and `[0x1005,0x1009]`), directly pinning the "split at unsupported instructions" claim rather than only the front/back-of-run cases.
Validation: new test passes; `cargo fmt --all -- --check` and the `candidate_windows` suite green.

**Non-4-byte-aligned AArch64 section edge case → Deferred, tracked as #677.**
Real but effectively unreachable on this discovery seam: all AArch64 executable content is emitted in 4-byte instruction units, so an instruction-only `.text` length is a multiple of 4 by construction, and if it were ever hit `validate_address_window` fails closed with a descriptive error rather than an unsound window. Filed #677 to confirm/handle it explicitly rather than expand this PR's scope.

**`#[cfg_attr(not(test), allow(dead_code))]` stopgaps → No change (already tracked).**
These are load-bearing right now — the seam is genuinely dead under `cfg(not(test))` until #620 wires it into the CLI driver, so removing them would break the `-D warnings` build. Their removal is already documented in-code against #620 and falls out naturally when that lands; no separate tracking needed.

_Triaged by two independent AI reviewers (Claude Opus 4.8 and Codex GPT-5.x). If you disagree on any of these, reply and we'll re-evaluate._

File: src/main.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main.rs]: Fixed in c313223 by replacing the slice iter().nth(1) lookup with direct get(1) access. Verified with cargo test --bin s11 candidate_window, cargo clippy --all -- -D warnings, cargo test --all, and ./ci_check.sh; all applicable checks pass.

**Comment by pmatos** [comment]
Triaged the three follow-up review points with two independent evaluators (Claude Opus 4.8 and Codex GPT-5.x) — all three are polish/verification items, no code change:

1. **`#[cfg_attr(not(test), allow(dead_code))]` placement → verified clean, no change.** `find_candidate_windows_with_backend` already carries the attribute; only the two `capstone_detail_*` helpers don't, and they're reachable both from that allow-marked caller and from tests, so they never warn in either config. Confirmed empirically: the green `clippy` job (`cargo clippy --all -- -D warnings`, compiled under `cfg(not(test))`) and a local plain `cargo build` are both warning-free — the non-test build has no dead-code warnings. Adding redundant attributes would just be churn.

2. **Double iteration over instructions → no change.** The two passes are intentionally distinct: the `try_fold` is a whole-section fail-closed decode-integrity gate (`ensure_window_fully_decoded_for_arch`) that must complete *before* any window is trusted, and the second pass builds windows. Folding them would entangle validation with construction and weaken the fail-closed semantics for no measurable gain on not-yet-wired scaffolding.

3. **AArch64 section alignment edge case → already tracked in #677.** Same point raised in the earlier review; deferred and filed as #677 rather than duplicated here.

_Triaged by two independent AI reviewers (Claude Opus 4.8 and Codex GPT-5.x)._
